tensorboard/components/vz_line_chart2/line-chart-exporter.ts

```diff
@@ -12,136 +12,126 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-        const style = window.getComputedStyle(node as Element);
-        const left = parseInt(style.left, 10);
-        const top = parseInt(style.top, 10);
-        if (left || top) {
-          const clipId = this.createUniqueId('clip');
+export class PlottableExporter {
+  private root: Element;
+  private uniqueId: number = 0;
+  constructor(rootEl: Element) {
+    this.root = rootEl;
+  }
+  public exportAsString(): string {
+    const convertedNodes = this.convert(this.root);
+    if (!convertedNodes) return '';
+    const svg = this.createRootSvg();
+    svg.appendChild(convertedNodes);
+    return svg.outerHTML;
+  }
+  private createUniqueId(prefix: string): string {
+    return `${prefix}_${this.uniqueId++}`;
+  }
+  private getSize(): DOMRect | ClientRect {
+    return this.root.getBoundingClientRect();
+  }
+  private createRootSvg(): Element {
+    const svg = document.createElement('svg');
+    const rect = this.getSize();
+    // case on `viewBox` is sensitive.
+    svg.setAttributeNS('svg', 'viewBox', `0 0 ${rect.width} ${rect.height}`);
+    svg.setAttribute('xmlns', 'http://www.w3.org/2000/svg');
+    return svg;
+  }
+  private convert(node: Node): Node | null {
+    let newNode: Node | null = null;
+    const nodeName = node.nodeName.toUpperCase();
+    if (
+      node.nodeType == Node.ELEMENT_NODE &&
+      (nodeName == NodeName.DIV || nodeName == NodeName.SVG)
+    ) {
+      newNode = document.createElement(NodeName.GROUP);
+      const style = window.getComputedStyle(node as Element);
+      const left = parseInt(style.left || '', 10);
+      const top = parseInt(style.top || '', 10);
+      if (left || top) {
+        const clipId = this.createUniqueId('clip');
+        if (newNode instanceof Element) {
           newNode.setAttribute('transform', `translate(${left}, ${top})`);
           newNode.setAttribute('clip-path', `url(#${clipId})`);
-          const width = parseInt(style.width, 10);
-          const height = parseInt(style.height, 10);
-          const rect = document.createElement('rect');
-          rect.setAttribute('width', String(width));
-          rect.setAttribute('height', String(height));
-          const clipPath = document.createElementNS('svg', 'clipPath');
-          clipPath.id = clipId;
-          clipPath.appendChild(rect);
-          newNode.appendChild(clipPath);
         }
-      } else {
-        newNode = node.cloneNode();
+        const width = parseInt(style.width || '', 10);
+        const height = parseInt(style.height || '', 10);
+        const rect = document.createElement('rect');
+        rect.setAttribute('width', String(width));
+        rect.setAttribute('height', String(height));
+        const clipPath = document.createElementNS('svg', 'clipPath');
+        clipPath.id = clipId;
+        clipPath.appendChild(rect);
+        newNode.appendChild(clipPath);
       }
+    } else {
+      newNode = node.cloneNode();
+    }
+
+    if (newNode) {
```

Comment:
Aren't we guaranteed newNode is valid by here?

Reply:
Right, I think I meant to write `!` which should achieve the same effect (otherwise, it is nullable)